map/client/client.go

```diff
@@ -0,0 +1,335 @@
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"github.com/fxamacker/cbor/v2"
+	"github.com/katzenpost/katzenpost/client"
+	"github.com/katzenpost/katzenpost/client/utils"
+	"github.com/katzenpost/katzenpost/core/crypto/eddsa"
+	"github.com/katzenpost/katzenpost/map/common"
+	"golang.org/x/crypto/hkdf"
+	"sort"
+)
+
+var (
+	cborFrameOverhead = 0
+	hash              = sha256.New
+	ErrStatusNotFound = errors.New("StatusNotFound")
+)
+
+type Client struct {
+	Session *client.Session
+}
+
+func NewClient(s *client.Session) (*Client, error) {
+	return &Client{Session: s}, nil
+}
+
+type StorageLocation interface {
+	ID() common.MessageID
+	Name() string
+	Provider() string
+}
+
+type mapStorage struct {
+	id             common.MessageID
+	name, provider string
+}
+
+func (m *mapStorage) ID() common.MessageID {
+	return m.id
+}
+func (m *mapStorage) Name() string {
+	return m.name
+}
+func (m *mapStorage) Provider() string {
+	return m.provider
+}
+
+type DeterministicDescriptorList []utils.ServiceDescriptor
+
+func (d DeterministicDescriptorList) Less(i, j int) bool {
+	return d[i].Name+d[i].Provider < d[j].Name+d[j].Provider
+}
+func (d DeterministicDescriptorList) Len() int {
+	return len(d)
+}
+func (d DeterministicDescriptorList) Swap(i, j int) {
+	d[i], d[j] = d[j], d[i]
+}
+
+func deterministicSelect(descs []utils.ServiceDescriptor, slot int) utils.ServiceDescriptor {
+	// TODO: order the descriptors by their identity keys
+	ddescs := DeterministicDescriptorList(descs)
+	sort.Sort(ddescs)
+	return ddescs[slot]
+}
+
+// GetStorageProvider returns the deterministically selected storage provider given a storage secret ID
+func (c *Client) GetStorageProvider(ID common.MessageID) (StorageLocation, error) {
+	// doc must be current document!
+	doc := c.Session.CurrentDocument()
+	if doc == nil {
+		return nil, errors.New("No PKI document") // XXX: find correct error
+	}
+	descs := utils.FindServices(common.MapServiceName, doc)
+	if len(descs) == 0 {
+		return nil, errors.New("No descriptors")
+	}
+	slot := int(binary.LittleEndian.Uint64(ID[:8])) % len(descs)
+	// sort the descs and return the chosen one
+	desc := deterministicSelect(descs, slot)
+	return &mapStorage{name: desc.Name, provider: desc.Provider, id: ID}, nil
+}
+
+// Put places a value into the store
+func (c *Client) Put(ID common.MessageID, signature, payload []byte) error {
+	if !ID.WritePk().Verify(signature, payload) {
```

Comment:
Client's Put method here is mixing IO with cryptography which would normally be totally acceptable if the core protocol actually had the call to the cryptographic verification. as it stands now, this abstraction is difficult to test.